numpy/core/include/numpy/_dtype_api.h

```diff
@@ -199,6 +199,14 @@ typedef int (get_reduction_initial_function)(
         PyArrayMethod_Context *context, npy_bool reduction_is_empty,
         char *initial);
 
+/**
+ * A function to fill an array with an initial value.
+ *
+ * @param arr The array to be filled.
+ *
+ * @returns -1 or 0 indicating error and arr being successfully filled.
+ */
+typedef int (fill_initial_function)(PyArrayObject *arr);
```

Comment:
OK, fair enough! I think it might be possible to write this using a traversal loop. I will poke around next week.

Reply:
Cool, thanks.  I don't have any opinion right now about whats better.  The first is neat, but the second seems simpler and close to how identities currently work...

Reply:
Where in numpy is the ufunc identity buffer setup? I’ll take a look at that first since I’m not familiar with it.

Reply:
Its the `get_reduction_initial_function  *get_reduction_initial;` and defined here: https://github.com/numpy/numpy/blob/fd50e7f155d4c3313c71f60ab8af9bc4b2e6e354/numpy/core/include/numpy/_dtype_api.h#L178-L200

The actual implementation is more annoying, since its set up to fall back to the one on the ufunc: https://github.com/numpy/numpy/blob/1bbe5821cc6afa134221367640380f403fa8d011/numpy/core/src/umath/legacy_array_method.c#L266

(which is all a bit annoying and I wouldn't mind making it nicer, but it helps support complicated reduce initials.)

---

But of course what we would need here is much simpler, since it only needs the descr.

Reply:
I agree with not having things like this on the dtype. Having a `ufunc` would seem to make a lot of sense to me. Could be an actual `zeros(out=array`, or be more like the strange private `_ones_like` that already exists (no idea why! but it is covered by astropy units...). For zero filling it would then be `zeros_like(array, out=array)`.  But another logical concept is that for dtypes that do not have strange non-identical zeros, one effectively just uses whatever `np.copyto` does.

(Sorry, I have *not* really looked at this PR, just really read the comments here... Ignore if this is all irrelevant.)

Reply:
I had thought about a ufunc, but overall still lean against it.

My thoughts are roughly:
* Creating a zero filled array seems very core and doing ufunc dispatching is a bit of work for something simple.  It *feels* a bit much if I think of an exposing API like `PyArray_FillBuffer(dtype, ptr, num, stride, NPY_ZEROVAL)`  (that is partially a lie, because we made sure ufuncs are now set up in a way to make exactly this fairly easy!).
* Its just seems like more hassle for the DType author to implement for something pretty basic (but of course users will implement many ufuncs anyway).

Another path would be an `ArrayMethod` (i.e. ufunc loop) that lives on the DType (just like casts are ufuncs loops/ArrayMethods that don't live on a ufunc object).

For clearing (dealloc+NULLint of array values) I had for a simplified traversal loop getter.  I don't like splitting from `ArrayMethod`, but it seemed like a lot of set up for something very basic.  I do translate this to `zero()`.

Now ufuncs seem like they should have some advantages (maybe for subclasses?), but I can't really put my finger on anything concrete right now except a certain beauty of aiming that as much as possible should a ufunc.

---

N.B.: I suspect this will continue to include "zero", "one", and also "empty" (we do need this for `object()` although I am not sure we will need it generally).
In principle one could continue down the rabbit hole with the typical `ufunc.identity` values, and `NA` filling might be a thing at some point.

Reply:
Yeah, it is no more than thinking that there is a certain beauty and simplicity if a `DType` interacts with arrays only via ufuncs. Of course, regardless the `DType` has to have a way to tell what "zero" is (though there is an obvious default).